### PR TITLE
thunderbird calendar tasks module

### PR DESCRIPTION
**Pull request by mrt-prodz**
This module displays Thunderbird Calendar tasks:

>    {due} : due tasks
>    {completed} : completed tasks
>    {current} : title of current running task (sorted by priority and stamp)

Caution, in order to work user must set **profile_path** and enter the full path of his/her thunderbird profile folder in the i3satus config:

> ...
> order += "thunderbird_calendar"
> ...
> thunderbird_calendar {
>     profile_path = "/home/user/.thunderbird/1yawevtp.default"
>     format = "✔ {due} - {current}"
> }
> ...

Any input would be greatly appreciated.

### CODE REVIEW

**Comment by pydsigner** [comment]
Made a PR on your branch to adjust a few things. Your code looked like it did the right thing, but not in the right way.

**Comment by mrt-prodz** [comment]
Thank you, it's much cleaner that way.

**Comment by pydsigner** [comment]
Now that I think of it though, selecting `task[0]` in the for loop would be cleaner yet.

File: py3status/modules/thunderbird_calendar.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+"""
+Display tasks in thunderbird calendar.
+
+Configuration parameters:
+    - cache_timeout : how often we refresh usage in seconds (default: 120s)
+    - profile_path : path to the user thunderbird profile (not optional)
+    - format : see placeholders below
```
Comment (tobes) [inline py3status/modules/thunderbird_calendar.py]: Could you change `- cache_timeout : ` to `cache_timeout:`

This is for the new documentation.  I really should write a guide to the docstrings explaining this

**Comment by mrt-prodz** [comment]
Like as the following?

`duetasks = [task[0] for task in tasks if task[1] is None]`

And then returning:

`current = duetasks[0]`

**Comment by pydsigner** [comment]
Exactly.

File: py3status/modules/thunderbird_calendar.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+Format of status string placeholders:
+    {due} : due tasks
+    {completed} : completed tasks
+    {current} : title of current running task (sorted by priority and stamp)
```
Comment (tobes) [inline py3status/modules/thunderbird_calendar.py]: again `{current} :` to `{current}:`

File: py3status/modules/thunderbird_calendar.py (hunk 1 above)
Comment (mrt-prodz) [inline py3status/modules/thunderbird_calendar.py]: Should I remove any space after configuration parameter?

File: py3status/modules/thunderbird_calendar.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+from sqlite3 import connect
+from os import access, R_OK
+from time import time
+
+
+class Py3status:
+    # available configuration parameters
+    cache_timeout = 120
+    # user must configure the path to thunderbird profile_path
+    # ex: /home/user/.thunderbird/1yawevtp.default
```
Comment (tobes) [inline py3status/modules/thunderbird_calendar.py]: This is in the docstring so probably doesn't need repeating here.

File: py3status/modules/thunderbird_calendar.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        }
+        return response
+
+    # return calendar data
+    def get_calendar(self, i3s_output_list, i3s_config):
+        db = self.profile_path + '/calendar-data/local.sqlite'
+        due = completed = 0
+        current = ''
+        if not access(db, R_OK):
+            return self._error_response(i3s_config['color_bad'])
```
Comment (tobes) [inline py3status/modules/thunderbird_calendar.py]: Maybe you could pass to the user a nice error eg `No access - check profile_path`

**Comment by mrt-prodz** [comment]
@tobes would the documentation be following guidelines as such:

> Configuration parameters:
>     cache_timeout: how often we refresh usage in seconds (default: 120s)
>     profile_path: path to the user thunderbird profile (not optional)
>     format: see placeholders below
 
> Format of status string placeholders:
>     {due} due tasks
>     {completed} completed tasks
>     {current} title of current running task (sorted by priority and stamp)

Configuration parameters shouldn't have a dash nor a space between the var name and the colon, and format placeholders shouldn't have any symbol either. Is that correct?

Checking other modules I see that no dash or colon should be after the format placeholder either.

File: py3status/modules/thunderbird_calendar.py (hunk 1 above)
Comment (tobes) [inline py3status/modules/thunderbird_calendar.py]: yes,

if you look in master branch all the modules now conform so you can check with them.  They get transformed for the README.md

basically
```
One line desc

general blurb

Configuration parameters:
    cache_timeout: how often we refresh usage in seconds (default: 120s)
    profile_path: path to the user thunderbird profile 

Format of status string placeholders:
    {due}: due tasks
    {completed}: completed tasks

@author me
```

File: py3status/modules/thunderbird_calendar.py (hunk 2 above)
Comment (tobes) [inline py3status/modules/thunderbird_calendar.py]: actualy I'm wrong should be `{current} desc` no colons

**Comment by tobes** [comment]
````
Configuration parameters:
    cache_timeout: how often we refresh usage in seconds (default: 120s)
    profile_path: path to the user thunderbird profile (not optional)
    format: see placeholders below

Format of status string placeholders:
    {due} due tasks
    {completed} completed tasks
    {current} title of current running task (sorted by priority and stamp)
```

they indent

**Comment by mrt-prodz** [comment]
@tobes you are correct, pointing to the user a possible issue with profile_path would be much better than only outputting Err.

Would `profile_path | read permission error` and `calendar parsing error` be more appropriate?
(the later would be if any exception has been raised)

**Comment by pydsigner** [comment]
The error message could be customizable

**Comment by mrt-prodz** [comment]
For now I just added a text parameter to _error_response, we could allow two configurable variables for the profile_path and exception error. It could be annoying for users to have a big error message, especially if the status is already showing a lot of information.

**Comment by tobes** [comment]
if you run `py3status docstring update thunderbird_calendar` it should update  `modules/README.md`
assuming you are running you branch as py3status

File: py3status/modules/thunderbird_calendar.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+"""
+from sqlite3 import connect
+from os import access, R_OK
+from time import time
+
+
+class Py3status:
+    # available configuration parameters
+    cache_timeout = 120
+    profile_path = ''
```
Comment (ultrabug) [inline py3status/modules/thunderbird_calendar.py]: please sort the variables alphabetically (same on docstring above)

Comment (mrt-prodz) [inline py3status/modules/thunderbird_calendar.py]: Should it be like this?

>Configuration parameters:
>&nbsp;&nbsp;&nbsp;&nbsp;cache_timeout: how often we refresh usage in seconds (default: 120s)
>&nbsp;&nbsp;&nbsp;&nbsp;err_exception: error message when an exception is raised
>&nbsp;&nbsp;&nbsp;&nbsp;err_profile: error message regarding profile path and read access
>&nbsp;&nbsp;&nbsp;&nbsp;format: see placeholders below
>&nbsp;&nbsp;&nbsp;&nbsp;profile_path: path to the user thunderbird profile (not optional)
>
>Format of status string placeholders:
>&nbsp;&nbsp;&nbsp;&nbsp;{completed} completed tasks
>&nbsp;&nbsp;&nbsp;&nbsp;{current} title of current running task (sorted by priority and stamp)
>&nbsp;&nbsp;&nbsp;&nbsp;{due} due tasks
>

...

>class Py3status: 
>&nbsp;&nbsp;&nbsp;&nbsp;# available configuration parameters
>&nbsp;&nbsp;&nbsp;&nbsp;cache_timeout = 120
>&nbsp;&nbsp;&nbsp;&nbsp;err_exception = 'error: calendar parsing failed'
>&nbsp;&nbsp;&nbsp;&nbsp;err_profile = 'error: profile not readable'
>&nbsp;&nbsp;&nbsp;&nbsp;format = 'tasks:[{due}] current:{current}'
>&nbsp;&nbsp;&nbsp;&nbsp;profile_path = ''
>

Comment (tobes) [inline py3status/modules/thunderbird_calendar.py]: looks good

**Comment by raspbeguy** [comment]
Since this PR seems to be stalled, I thing I'm gonna work on a more global calendar/tasks module that could check on thunderbird, but also directly from caldav.

**Comment by mrt-prodz** [comment]
I'm really not sure where it stalled, I actually use this module every day since this PR. Didn't encounter any issue so far.

**Comment by tobes** [comment]
> I'm really not sure where it stalled

Me neither.  Anyhow things have change since this was created so it would need a slight refresh

* documentation defaults

* `i3s_config['color_bad']` -> `self.py3.COLOR_BAD`

* use `self.py3.safe_format()`

* new standalone testing

**Comment by lasers** [comment]
Hi all, I played with this module. This works fine with preview release version `57.0 beta`. My opinion is that this module could be closed instead for few reasons.

* No word from the author.
* Module is named `thunderbird_calendar` which is a misleading (imo) as it does not deal with calendar. Just tasks.
* An issue of making a selected task `current` so it can show up on the bar. There is a criteria to make it current, but it is not clear to the users and/or does not always work.
* Hardcoded criteria `(sorted by priority and stamp)` -- does not always work.
* This is not done in the loop so it does not display all tasks.
* Module too simple and/or too limited.

Semi-related: I also wrote `taskwarrior` module from scratch that can do much much more and the tool `taskwarrior` itself seems to be an excellent `task` tool to use/recommend. Thx.

**Comment by lasers** [comment]
@ultrabug @tobes: Address this PR, please.

**Comment by ultrabug** [comment]
Stall comes from me I guess and I apologize for this @mrt-prodz 

TBH, author did a good job and I'm the one to blame here. Thunderbird users may benefit from it and all that what @tobes address can be fixed by us after the merge so that @mrt-prodz keeps the logic contribution of the module.

If nothing new comes against this in the following days, I'll merge & fix afterwards.

**Comment by lasers** [comment]
Start with this, I guess. https://github.com/lasers/py3status/commit/294a6d992038b4c57b6f975a0afc526d805fbeb2

**Comment by ultrabug** [comment]
@lasers I'd rather merge this for authorship respect and then you doing a PR to fix/enhance the module as co author if you may?

**Comment by lasers** [comment]
Sure. Sounds good.

**Comment by mrt-prodz** [comment]
@ultrabug there is no need to apologize, I should have followed through but I got caught up with too many things since then. (moved country, got a new job keeping me really busy etc..)

I personally used this for quite some time until I decided to upgrade my system with the new py3status version. I agree with @lasers that many things could be done to improve this module and that the name might not really reflect on what it really does.

**Comment by lasers** [comment]
@mrt-prodz Hi. Can you give this module a try? Let me know if you like it and please do share your concerns (if any). I think it's basically done when we fill in the blanks. Thanks. [thunderbirds_todos.py](https://github.com/lasers/py3status/blob/newmodule-thunderbird-todos/py3status/modules/thunderbird_todos.py)

**Comment by ultrabug** [comment]
@mrt-prodz and @lasers what do you want me to do then? I'd merge this only if there's at least a real user behind it, else maybe it's too old and it isn't worth adding to the code base?

Thanks

**Comment by lasers** [comment]
@ultrabug...

>@mrt-prodz I personally used this for quite some time until I decided to upgrade my system with the new py3status version.

It's up to him (a real user). You can merge this and I put out a new pull request that can do more: Display all todos, work with boolean, thresholds, converting datetimes, and whatnot.

If he does not like my module, he can use his... and I think my module is able to similarily reproduce his behavior with little changes. I want to know first if he is satisfied with `[\?if=index=1 {todo}]` to display only the first todo? or do he explicitly need this...
```
            cur.execute('SELECT title, todo_completed FROM cal_todos '
                        'ORDER BY priority DESC, todo_stamp DESC')
```
I use...
```
            cur.execute('SELECT * FROM cal_todos')
```

@mrt-prodz's module works fine with latest thunderbird beta, is too simple, and was made for him with `ORDER BY` code above. The updated module makes it much more usable for all users. Hopefully @mrt-prodz too. Also, we're talking about todos. Not much to go on with this.

**Comment by mrt-prodz** [comment]
@lasers @ultrabug I'm testing this as we speak, @lasers you did a great job.

Right now ```format = '\?color=incompleted {incompleted} todos'``` is working just fine but I'm struggling with displaying the title correctly using ```format_todo```. I guess I'm doing something wrong in the configuration, need to figure what is incorrect exactly.

I'm following the examples in the module description but it's returning ```unhashable type: 'list'```.

**Comment by lasers** [comment]
@mrt-prodz I pushed out a `troubleshooting` commit. Start from there and track down the issue... Maybe share the output too. I'm assuming this had to do more with me using barebone data and you using real data.

**Comment by ultrabug** [comment]
Ok I'm now merging this to retain authorship. Now plz @lasers move from this and make a proper PR to pass in your improvements thanks!

@mrt-prodz thanks again, I hope you'll help @lasers in his continued PR on this module 👍

**Comment by mrt-prodz** [comment]
@lasers found where the issue is actually.

https://github.com/lasers/py3status/blob/newmodule-thunderbird-todos/py3status/modules/thunderbird_todos.py#L154

```not self.py3.format_contains(self.format, ['todo', 'incompleted']) and \```

should be

```not self.py3.format_contains(self.format, ('todo', 'incompleted')) and \```

**Comment by mrt-prodz** [comment]
@lasers Trying to figure what is going on exactly, I quickly looked over ```py3.format_contains``` and the syntax should be working (the one you wrote). It's been a while since the last time I worked on a module for py3status and many things changed, but maybe we are looking in the right direction.

**Comment by mrt-prodz** [comment]
When running the module directly to test things:

```
Exception in `test_module` post_config_hook() (TypeError) py3.py line 614.
Traceback
TypeError: unhashable type: 'list'
  File "/usr/lib/python2.7/site-packages/py3status/module.py", line 137, in prepare_module
    self.module_class.post_config_hook()
  File "thunderbird_todos.py", line 153, in post_config_hook
    not self.py3.format_contains(self.format, ['todo', 'incompleted']) and \
  File "/usr/lib/python2.7/site-packages/py3status/py3.py", line 614, in format_contains
    return self._format_placeholders_cache[format_string][name]

terminating module test_module
```
*As a side note I'm using py3status v3.6 from Fedora 26 repository.*

**Comment by lasers** [comment]
>As a side note I'm using py3status v3.6 from Fedora 26 repository.

That's the issue.

**Comment by mrt-prodz** [comment]
Good to know, before making you waste your time with this I wanted to make things clear that I'm not up to date with upstream. I can see modules like diskdata.py using this format as well.

**Comment by mrt-prodz** [comment]
After upgrading to py3status v3.7, it's now working just fine.

```format_todo = '[\?if=index=1 {title}]'``` is actually doing what I wanted in the original module, so it's working just fine for me for my specific use.

**Pull request merged**